File: src/hasoftware-library/src/hasoftware/api/messages/InputMessageResponse.java
package hasoftware.api.messages;

import hasoftware.api.FunctionCode;
import hasoftware.api.Message;
import hasoftware.api.classes.InputMessage;
import hasoftware.cdef.CDEFAction;
import hasoftware.cdef.CDEFMessage;
import hasoftware.cdef.CDEFSystemFlags;
import java.util.LinkedList;
import java.util.List;

public class InputMessageResponse extends Message {

    private int _action;
    private final List<InputMessage> _inputMessages = new LinkedList<>();

    public InputMessageResponse(int transactionNumber) {
        super(FunctionCode.InputMessage, transactionNumber, CDEFSystemFlags.Response);
        _action = CDEFAction.None;
    }

    public InputMessageResponse(CDEFMessage cdefMessage) {
        super(cdefMessage);
        _action = cdefMessage.getU8();
        int countInputMessages = cdefMessage.getU8();
        for (int index = 0; index < countInputMessages; index++) {
            _inputMessages.add(new InputMessage(cdefMessage));
        }
    }

    @Override
    public void encode(CDEFMessage cdefMessage) {
        super.encode(cdefMessage);
        cdefMessage.putU8(_action);
        cdefMessage.putU8(_inputMessages.size());
        for (InputMessage inputMessage : _inputMessages) {
            inputMessage.encode(cdefMessage);
        }
    }

    public int getAction() {
        return _action;
    }

    public void setAction(int action) {
        _action = action;
    }

    public List<InputMessage> getInputMessages() {
        return _inputMessages;
    }
}

File: src.net/hasoftware-library/Messages/InputMessageResponse.cs
using hasoftware.Api;
using hasoftware.Cdef;
using hasoftware.Classes;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace hasoftware.Messages
{
    public class InputMessageResponse : Message
    {
        public CdefAction Action { get; private set; }
        public List<InputMessage> InputMessages { get; private set; }

        public InputMessageResponse(int transactionNumber)
            :base(Api.FunctionCode.InputMessage, transactionNumber, CdefSystemFlags.Response)
        {
            Action = CdefAction.None;
            InputMessages = new List<InputMessage>();
        }

        public InputMessageResponse(CdefMessage cdefMessage)
            :base(cdefMessage)
        {
            Action = (CdefAction)cdefMessage.GetU8();
            int countInputMessages = cdefMessage.GetU8();
            for (int index = 0; index < countInputMessages; index++) {
                InputMessages.Add(new InputMessage(cdefMessage));
            }
        }

        public override void Encode(CdefMessage cdefMessage) {
            base.Encode(cdefMessage);
            cdefMessage.PutU8((int)Action);
            cdefMessage.PutU8(InputMessages.Count);
            foreach (var inputMessage in InputMessages) {
                inputMessage.Encode(cdefMessage);
            }
        }
    }
}
